pkg/scale/encode.go

```diff
@@ -0,0 +1,420 @@
+package scale
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"math/big"
+	"reflect"
+)
+
+func Marshal(v interface{}) (b []byte, err error) {
+	es := encodeState{
+		fieldScaleIndicesCache: cache,
+	}
+	err = es.marshal(v)
+	if err != nil {
+		return
+	}
+	b = es.Bytes()
+	return
+}
+
+type encodeState struct {
+	bytes.Buffer
+	*fieldScaleIndicesCache
+}
+
+func (es *encodeState) marshal(in interface{}) (err error) {
+	switch in := in.(type) {
+	case int:
+		err = es.encodeUint(uint(in))
+	case uint:
+		err = es.encodeUint(in)
+	case int8, uint8, int16, uint16, int32, uint32, int64, uint64:
+		err = es.encodeFixedWidthInt(in)
+	case *big.Int:
+		err = es.encodeBigInt(in)
+	case *Uint128:
+		err = es.encodeUint128(in)
+	case []byte:
+		err = es.encodeBytes(in)
+	case string:
+		err = es.encodeBytes([]byte(in))
+	case bool:
+		err = es.encodeBool(in)
+	default:
+		switch reflect.TypeOf(in).Kind() {
+		case reflect.Ptr:
+			// Assuming that anything that is a pointer is an Option to capture {nil, T}
+			elem := reflect.ValueOf(in).Elem()
+			switch elem.IsValid() {
+			case false:
+				err = es.WriteByte(0)
+			default:
+				err = es.WriteByte(1)
+				if err != nil {
+					return
+				}
+				err = es.marshal(elem.Interface())
+			}
+		case reflect.Struct:
+			t := reflect.TypeOf(in)
+			// check if this is a convertible to VaryingDataType, if so encode using encodeVaryingDataType
+			switch t.ConvertibleTo(reflect.TypeOf(Result{})) {
+			case true:
+				resv := reflect.ValueOf(in).Convert(reflect.TypeOf(Result{}))
+				err = es.encodeResult(resv.Interface().(Result))
+			case false:
+				err = es.encodeStruct(in)
+			}
+
+		case reflect.Array:
+			err = es.encodeArray(in)
+		case reflect.Slice:
+			t := reflect.TypeOf(in)
+			// check if this is a convertible to VaryingDataType, if so encode using encodeVaryingDataType
+			switch t.ConvertibleTo(reflect.TypeOf(VaryingDataType{})) {
+			case true:
+				invdt := reflect.ValueOf(in).Convert(reflect.TypeOf(VaryingDataType{}))
+				err = es.encodeVaryingDataType(invdt.Interface().(VaryingDataType))
+			case false:
+				err = es.encodeSlice(in)
+			}
+		default:
+			_, ok := in.(VaryingDataTypeValue)
+			switch ok {
+			case true:
+				t := reflect.TypeOf(in)
+				switch t.Kind() {
+				// TODO: support more primitive types.  Do we need to support arrays and slices as well?
```

Comment:
same for decode, i think we'll need `common.Hash` which is `[32]byte` and `uint64`